# Data Processing

## Importing Data 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
os.getcwd()

'/Users/zora/Desktop/springboard/NYC TAXI/DATA WRANGLING'

In [2]:
files = os.listdir()
print(files)

['.DS_Store', 'Data Processing.ipynb', '.ipynb_checkpoints', 'NYC TAXI DATA']


In [3]:
path='/Users/zora/Desktop/springboard/NYC TAXI/DATA WRANGLING/NYC TAXI DATA'
os.chdir(path) 

### Yellow Taxi

In [4]:
yellow_taxi = pd.read_csv('yellow_tripdata_2019-09.csv', index_col=False)
yellow_taxi.head()

/Users/zora/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3063: DtypeWarning: Columns (6) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2019-09-01 00:06:48   2019-09-01 00:25:46              1.0   
1       1.0  2019-09-01 00:29:17   2019-09-01 00:57:54              1.0   
2       1.0  2019-09-01 00:41:32   2019-09-01 00:41:36              1.0   
3       1.0  2019-09-01 00:59:11   2019-09-01 00:59:15              1.0   
4       1.0  2019-09-01 00:39:41   2019-09-01 00:45:06              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            2.0         1.0                  N           186           161   
1            5.5         1.0                  N           161           256   
2            2.9         1.0                  N           145           145   
3            0.0         1.0                  N           145           145   
4            1.5         1.0                  N            95            93   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0           2.0         13.0    3.0      0.5         0.0           0.0   
1           2.0         22.5    3.0      0.5         0.0           0.0   
2           2.0          2.5    0.5      0.5         0.0           0.0   
3           2.0          2.5    0.5      0.5         0.0           0.0   
4           2.0          7.0    0.5      0.5         0.0           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  
0                    0.3          16.8                   2.5  
1                    0.3          26.3                   2.5  
2                    0.3           3.8                   0.0  
3                    0.3           3.8                   0.0  
4                    0.3           8.3                   0.0

In [5]:
col_names = list(yellow_taxi.columns)
print(col_names)
print(yellow_taxi.shape)
yellow_taxi.info()

['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime', 'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag', 'PULocationID', 'DOLocationID', 'payment_type', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge', 'total_amount', 'congestion_surcharge']
(6567788, 18)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6567788 entries, 0 to 6567787
Data columns (total 18 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   VendorID               float64
 1   tpep_pickup_datetime   object 
 2   tpep_dropoff_datetime  object 
 3   passenger_count        float64
 4   trip_distance          float64
 5   RatecodeID             float64
 6   store_and_fwd_flag     object 
 7   PULocationID           int64  
 8   DOLocationID           int64  
 9   payment_type           float64
 10  fare_amount            float64
 11  extra                  float64
 12  mta_tax                float64
 13  tip_amount      

In [6]:
yellow_taxi.describe().T

count        mean        std     min    25%  \
VendorID               6533699.0    1.648833   0.477360    1.00    1.0   
passenger_count        6533699.0    1.549407   1.192190    0.00    1.0   
trip_distance          6567788.0    3.086974   4.036779    0.00    1.0   
RatecodeID             6533699.0    1.064147   0.840010    1.00    1.0   
PULocationID           6567788.0  162.892143  66.202447    1.00  114.0   
DOLocationID           6567788.0  161.126577  70.503403    1.00  107.0   
payment_type           6533699.0    1.287302   0.480756    1.00    1.0   
fare_amount            6567788.0   13.812060  60.529408 -530.00    6.5   
extra                  6567788.0    1.150874   1.272248   -7.50    0.0   
mta_tax                6567788.0    0.493713   0.065979   -0.50    0.5   
tip_amount             6567788.0    2.277880   2.937932  -80.08    0.0   
tolls_amount           6567788.0    0.408465   1.853549  -40.00    0.0   
improvement_surcharge  6567788.0    0.298348   0.030410   -0.30    0.3   
total_amount           6567788.0   19.918379  61.192518 -530.80   11.6   
congestion_surcharge   6567787.0    2.275256   0.730913   -2.50    2.5   

                          50%     75%        max  
VendorID                 2.00    2.00       4.00  
passenger_count          1.00    2.00       9.00  
trip_distance            1.69    3.14     603.50  
RatecodeID               1.00    1.00      99.00  
PULocationID           162.00  233.00     265.00  
DOLocationID           162.00  233.00     265.00  
payment_type             1.00    2.00       4.00  
fare_amount              9.50   15.50  150998.39  
extra                    0.50    2.50      32.50  
mta_tax                  0.50    0.50      15.55  
tip_amount               1.95    3.00     496.20  
tolls_amount             0.00    0.00     900.09  
improvement_surcharge    0.30    0.30       1.00  
total_amount            15.30   21.36  151007.81  
congestion_surcharge     2.50    2.50       2.75

### Data Cleaning

### Handle the unexpected values

In [8]:
unique_values_for_cols = yellow_taxi.nunique()
print(unique_values_for_cols)

VendorID                       3
tpep_pickup_datetime     2129612
tpep_dropoff_datetime    2130945
passenger_count               10
trip_distance               4583
RatecodeID                     7
store_and_fwd_flag             2
PULocationID                 261
DOLocationID                 263
payment_type                   4
fare_amount                 5184
extra                         45
mta_tax                       11
tip_amount                  3662
tolls_amount                1046
improvement_surcharge          5
total_amount               12548
congestion_surcharge           7
dtype: int64


##### Handle the Congestion Surcharge
Trips in taxis and For-Hire Vehicles will be charged a surcharge for trips that start, end or pass through Congestion Zone -- Manhattan, south of 96th Street.

The congestion surcharge, 
$2.50 for non-shared trips in taxicabs
$2.75 for non-shared trips in For-Hire-Vehicles, including limousines, and Street-Hail Liveries (Green Taxis)
OR
$0.75 for shared-rides in any type of vehicle
No surcharge for Access-A-Ride, or any other MTA dispatched trip.

The numbers from congestion_surcharge column should only be "2.50", "2.75", "0.75" or "0".

In [9]:
print(yellow_taxi['congestion_surcharge'].value_counts())

 2.50    5989239
 0.00     566323
-2.50      12027
 2.75        110
 0.75         86
 1.00          1
 0.50          1
Name: congestion_surcharge, dtype: int64


In [51]:
print(yellow_taxi[yellow_taxi['congestion_surcharge'].astype(str).str.contains('-2.5')])

         VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
508           2.0  2019-09-01 00:30:38   2019-09-01 00:31:49              1.0   
1026          2.0  2019-09-01 00:10:47   2019-09-01 00:12:56              5.0   
1185          2.0  2019-09-01 00:18:38   2019-09-01 00:29:45              1.0   
1187          2.0  2019-09-01 00:37:09   2019-09-01 00:55:51              1.0   
2295          2.0  2019-09-01 00:21:12   2019-09-01 00:52:46              2.0   
...           ...                  ...                   ...              ...   
6529807       2.0  2019-09-30 23:17:22   2019-09-30 23:24:26              1.0   
6530854       2.0  2019-09-30 23:36:21   2019-09-30 23:49:21              1.0   
6531661       2.0  2019-09-30 23:32:52   2019-09-30 23:36:21              1.0   
6531753       2.0  2019-09-30 23:51:30   2019-09-30 23:53:55              1.0   
6533239       2.0  2019-09-30 23:37:40   2019-09-30 23:52:39              1.0   

         trip_distance  Rat

In [10]:
print(yellow_taxi['VendorID'].value_counts())

2.0    4239200
1.0    2294473
4.0         26
Name: VendorID, dtype: int64


In [11]:
#print(yellow_taxi[yellow_taxi['VendorID'].astype(str).str.contains('4.0')])
yellow_taxi = yellow_taxi[yellow_taxi.VendorID != 4.0]
print(yellow_taxi['VendorID'].value_counts())

2.0    4239200
1.0    2294473
Name: VendorID, dtype: int64


##### Handle the missing and NA values

In [12]:
yellow_taxi.isnull().sum()

VendorID                 34089
tpep_pickup_datetime         0
tpep_dropoff_datetime        0
passenger_count          34089
trip_distance                0
RatecodeID               34089
store_and_fwd_flag       34089
PULocationID                 0
DOLocationID                 0
payment_type             34089
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
improvement_surcharge        0
total_amount                 0
congestion_surcharge         1
dtype: int64

In [13]:
nas=pd.DataFrame(yellow_taxi.isnull().sum().sort_values(ascending=False)/len(yellow_taxi)*100,columns = ['percent'])
pos = nas['percent'] > 0
nas[pos]

percent
VendorID              0.519035
payment_type          0.519035
passenger_count       0.519035
RatecodeID            0.519035
store_and_fwd_flag    0.519035
congestion_surcharge  0.000015

In [14]:
null_yellow = yellow_taxi[yellow_taxi.isna().any(axis=1)]
print (null_yellow)

         VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
4021083       2.0  2019-09-19 15:49:19   2019-09-19 15:49:41              2.0   
6533699       NaN  2019-09-01 07:51:00   2019-09-01 08:12:00              NaN   
6533700       NaN  2019-09-01 07:06:00   2019-09-01 07:23:00              NaN   
6533701       NaN  2019-09-01 07:07:00   2019-09-01 07:31:00              NaN   
6533702       NaN  2019-09-01 07:26:00   2019-09-01 07:49:00              NaN   
...           ...                  ...                   ...              ...   
6567783       NaN  2019-09-30 08:02:00   2019-09-30 08:20:00              NaN   
6567784       NaN  2019-09-30 08:00:00   2019-09-30 08:30:00              NaN   
6567785       NaN  2019-09-30 08:10:10   2019-09-30 08:46:23              NaN   
6567786       NaN  2019-09-30 08:17:00   2019-09-30 08:42:00              NaN   
6567787       NaN  2019-09-30 08:11:00   2019-09-30 08:26:00              NaN   

         trip_distance  Rat

In [15]:
yellow_df = yellow_taxi.dropna(axis = 0, how ='any')#drop 'congestion_surcharge' because 'trip_distance=0' as well.
  
# comparing sizes of data frames 
print("Old data frame length:", len(yellow_taxi), "\nNew data frame length:",  
       len(yellow_df), "\nNumber of rows with at least 1 NA value: ", 
       (len(yellow_taxi)-len(yellow_df)))


Old data frame length: 6567762 
New data frame length: 6533672 
Number of rows with at least 1 NA value:  34090


In [16]:
duplicateRowsDF = yellow_df[yellow_df.duplicated()]
duplicateRowsDF

Empty DataFrame
Columns: [VendorID, tpep_pickup_datetime, tpep_dropoff_datetime, passenger_count, trip_distance, RatecodeID, store_and_fwd_flag, PULocationID, DOLocationID, payment_type, fare_amount, extra, mta_tax, tip_amount, tolls_amount, improvement_surcharge, total_amount, congestion_surcharge]
Index: []

#### Adding EDA analysis label

In [17]:
yellow_df['Type'] = 'yellow'
yellow_df.head()

/Users/zora/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0       1.0  2019-09-01 00:06:48   2019-09-01 00:25:46              1.0   
1       1.0  2019-09-01 00:29:17   2019-09-01 00:57:54              1.0   
2       1.0  2019-09-01 00:41:32   2019-09-01 00:41:36              1.0   
3       1.0  2019-09-01 00:59:11   2019-09-01 00:59:15              1.0   
4       1.0  2019-09-01 00:39:41   2019-09-01 00:45:06              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            2.0         1.0                  N           186           161   
1            5.5         1.0                  N           161           256   
2            2.9         1.0                  N           145           145   
3            0.0         1.0                  N           145           145   
4            1.5         1.0                  N            95            93   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0           2.0         13.0    3.0      0.5         0.0           0.0   
1           2.0         22.5    3.0      0.5         0.0           0.0   
2           2.0          2.5    0.5      0.5         0.0           0.0   
3           2.0          2.5    0.5      0.5         0.0           0.0   
4           2.0          7.0    0.5      0.5         0.0           0.0   

   improvement_surcharge  total_amount  congestion_surcharge    Type  
0                    0.3          16.8                   2.5  yellow  
1                    0.3          26.3                   2.5  yellow  
2                    0.3           3.8                   0.0  yellow  
3                    0.3           3.8                   0.0  yellow  
4                    0.3           8.3                   0.0  yellow

### Green Taxi

In [18]:
green_taxi = pd.read_csv('green_tripdata_2019-09.csv', index_col=False)
green_taxi.head()

/Users/zora/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3063: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0       2.0  2019-09-01 00:10:53   2019-09-01 00:23:46                  N   
1       2.0  2019-09-01 00:31:22   2019-09-01 00:44:37                  N   
2       2.0  2019-09-01 00:50:24   2019-09-01 01:03:20                  N   
3       2.0  2019-09-01 00:27:06   2019-09-01 00:33:22                  N   
4       2.0  2019-09-01 00:43:23   2019-09-01 00:59:54                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            65           189              5.0           2.00   
1         1.0            97           225              5.0           3.20   
2         1.0            37            61              5.0           2.99   
3         1.0           145           112              1.0           1.73   
4         1.0           112           198              1.0           3.42   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0         10.5    0.5      0.5        2.36           0.0        NaN   
1         12.0    0.5      0.5        0.00           0.0        NaN   
2         12.0    0.5      0.5        0.00           0.0        NaN   
3          7.5    0.5      0.5        1.50           0.0        NaN   
4         14.0    0.5      0.5        3.06           0.0        NaN   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3         14.16           1.0        1.0   
1                    0.3         13.30           2.0        1.0   
2                    0.3         13.30           2.0        1.0   
3                    0.3         10.30           1.0        1.0   
4                    0.3         18.36           1.0        1.0   

   congestion_surcharge  
0                   0.0  
1                   0.0  
2                   0.0  
3                   0.0  
4                   0.0

In [19]:
col_names = list(green_taxi.columns)
print(col_names)
print(green_taxi.shape)
green_taxi.info()

['VendorID', 'lpep_pickup_datetime', 'lpep_dropoff_datetime', 'store_and_fwd_flag', 'RatecodeID', 'PULocationID', 'DOLocationID', 'passenger_count', 'trip_distance', 'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'ehail_fee', 'improvement_surcharge', 'total_amount', 'payment_type', 'trip_type', 'congestion_surcharge']
(449063, 20)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 449063 entries, 0 to 449062
Data columns (total 20 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   VendorID               389920 non-null  float64
 1   lpep_pickup_datetime   449063 non-null  object 
 2   lpep_dropoff_datetime  449063 non-null  object 
 3   store_and_fwd_flag     389920 non-null  object 
 4   RatecodeID             389920 non-null  float64
 5   PULocationID           449063 non-null  int64  
 6   DOLocationID           449063 non-null  int64  
 7   passenger_count        389920 non-null  float64
 8   tri

In [20]:
green_taxi.describe().T

count        mean        std     min   25%     50%  \
VendorID               389920.0    1.837777   0.368656    1.00   2.0    2.00   
RatecodeID             389920.0    1.106971   0.723302    1.00   1.0    1.00   
PULocationID           449063.0  108.139203  71.025032    1.00  56.0   82.00   
DOLocationID           449063.0  130.154678  76.577260    1.00  65.0  129.00   
passenger_count        389920.0    1.311900   0.969761    0.00   1.0    1.00   
trip_distance          449063.0    3.519991   4.083097    0.00   1.1    2.09   
fare_amount            449063.0   15.491455  13.561680 -200.00   7.0   11.00   
extra                  449063.0    0.697539   1.022319   -4.50   0.0    0.50   
mta_tax                449063.0    0.484626   0.097407   -0.50   0.5    0.50   
tip_amount             449063.0    1.037571   2.261574   -3.89   0.0    0.00   
tolls_amount           449063.0    0.312411   1.416886  -15.40   0.0    0.00   
ehail_fee                   0.0         NaN        NaN     NaN   NaN     NaN   
improvement_surcharge  449063.0    0.253791   0.110241   -0.30   0.3    0.30   
total_amount           449063.0   18.624612  15.243472 -200.00   8.8   13.55   
payment_type           389920.0    1.456383   0.522117    1.00   1.0    1.00   
trip_type              389920.0    1.024685   0.155162    1.00   1.0    1.00   
congestion_surcharge   389920.0    0.439440   1.007679   -2.75   0.0    0.00   

                          75%      max  
VendorID                 2.00     2.00  
RatecodeID               1.00    99.00  
PULocationID           166.00   265.00  
DOLocationID           195.00   265.00  
passenger_count          1.00     9.00  
trip_distance            4.22   341.64  
fare_amount             20.00  1761.00  
extra                    1.00     8.25  
mta_tax                  0.50    17.33  
tip_amount               1.66   400.00  
tolls_amount             0.00    33.24  
ehail_fee                 NaN      NaN  
improvement_surcharge    0.30     0.30  
total_amount            24.05  1762.80  
payment_type             2.00     5.00  
trip_type                1.00     2.00  
congestion_surcharge     0.00     2.75

In [21]:
unique_values_for_cols = green_taxi.nunique()
print(unique_values_for_cols)

VendorID                      2
lpep_pickup_datetime     375970
lpep_dropoff_datetime    375250
store_and_fwd_flag            2
RatecodeID                    7
PULocationID                257
DOLocationID                260
passenger_count              10
trip_distance              3142
fare_amount                4179
extra                        18
mta_tax                      16
tip_amount                 1683
tolls_amount                104
ehail_fee                     0
improvement_surcharge         3
total_amount               6175
payment_type                  5
trip_type                     2
congestion_surcharge          5
dtype: int64


#### Cleaning Data

##### Handle the Congestion Surcharge

##### Handle the Congestion Surcharge
Trips in taxis and For-Hire Vehicles will be charged a surcharge for trips that start, end or pass through Congestion Zone -- Manhattan, south of 96th Street.

The congestion surcharge, 
$2.50 for non-shared trips in taxicabs
$2.75 for non-shared trips in For-Hire-Vehicles, including limousines, and Street-Hail Liveries (Green Taxis)
OR
$0.75 for shared-rides in any type of vehicle
No surcharge for Access-A-Ride, or any other MTA dispatched trip.

The numbers from congestion_surcharge column should only be "2.50", "2.75", "0.75" or "0".

In [53]:
print(green_taxi['congestion_surcharge'].value_counts())

 0.00    327580
 2.75     62182
 2.50       146
-2.75         8
 0.75         4
Name: congestion_surcharge, dtype: int64


In [55]:
print(green_taxi[green_taxi['congestion_surcharge'].astype(str).str.contains('-')])

        VendorID lpep_pickup_datetime lpep_dropoff_datetime  \
91242        2.0  2019-09-08 04:09:34   2019-09-08 05:08:48   
172999       2.0  2019-09-14 12:50:29   2019-09-14 12:54:46   
237988       2.0  2019-09-19 14:46:30   2019-09-19 14:53:09   
241163       2.0  2019-09-19 18:21:43   2019-09-19 18:24:40   
322992       2.0  2019-09-25 19:51:54   2019-09-25 19:58:54   
339040       2.0  2019-09-26 22:04:08   2019-09-26 22:08:56   
345018       2.0  2019-09-27 13:17:37   2019-09-27 13:18:22   
361104       2.0  2019-09-28 15:38:08   2019-09-28 15:39:26   

       store_and_fwd_flag  RatecodeID  PULocationID  DOLocationID  \
91242                   N         1.0           255            87   
172999                  N         1.0            75           236   
237988                  N         1.0            75           263   
241163                  N         1.0            75           236   
322992                  N         1.0            43            43   
339040            

##### Handle the missing and NA values

In [22]:
green_taxi.isnull().sum()

VendorID                  59143
lpep_pickup_datetime          0
lpep_dropoff_datetime         0
store_and_fwd_flag        59143
RatecodeID                59143
PULocationID                  0
DOLocationID                  0
passenger_count           59143
trip_distance                 0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
ehail_fee                449063
improvement_surcharge         0
total_amount                  0
payment_type              59143
trip_type                 59143
congestion_surcharge      59143
dtype: int64

In [23]:
green_taxi['ehail_fee'] = green_taxi['ehail_fee'].fillna(0)
green_taxi['payment_type'] = green_taxi['payment_type'].fillna('5') #Unknown payment = 5

In [24]:
nas=pd.DataFrame(green_taxi.isnull().sum().sort_values(ascending=False)/len(green_taxi),columns = ['percent'])
pos = nas['percent'] > 0
nas[pos]

percent
congestion_surcharge  0.131703
trip_type             0.131703
store_and_fwd_flag    0.131703
RatecodeID            0.131703
passenger_count       0.131703
VendorID              0.131703

In [25]:
null_green = green_taxi[green_taxi.isna().any(axis=1)]
print (null_green.head())

        VendorID lpep_pickup_datetime lpep_dropoff_datetime  \
389920       NaN  2019-09-01 00:05:00   2019-09-01 00:39:00   
389921       NaN  2019-09-01 00:09:00   2019-09-01 00:25:00   
389922       NaN  2019-09-01 00:52:00   2019-09-01 01:24:00   
389923       NaN  2019-09-01 00:31:00   2019-09-01 01:11:00   
389924       NaN  2019-09-01 00:25:00   2019-09-01 01:14:00   

       store_and_fwd_flag  RatecodeID  PULocationID  DOLocationID  \
389920                NaN         NaN           191           117   
389921                NaN         NaN           235           213   
389922                NaN         NaN            61            90   
389923                NaN         NaN            61            74   
389924                NaN         NaN           119            89   

        passenger_count  trip_distance  fare_amount  extra  mta_tax  \
389920              NaN          12.45        45.95   2.75      0.5   
389921              NaN           5.11        29.25   2.75      

In [26]:
green_df = green_taxi.dropna(axis = 0, how ='any') 
  
# comparing sizes of data frames 
print("Old data frame length:", len(green_taxi), "\nNew data frame length:",  
       len(green_df), "\nNumber of rows with at least 1 NA value: ", 
       (len(green_taxi)-len(green_df)))

Old data frame length: 449063 
New data frame length: 389920 
Number of rows with at least 1 NA value:  59143


In [27]:
green_df['Type'] = 'green'
duplicateRows = green_df[green_df.duplicated()]
duplicateRows

/Users/zora/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


VendorID lpep_pickup_datetime lpep_dropoff_datetime  \
105063       2.0  2019-09-09 12:07:00   2019-09-09 12:07:00   

       store_and_fwd_flag  RatecodeID  PULocationID  DOLocationID  \
105063                  N         5.0           264           264   

        passenger_count  trip_distance  fare_amount  ...  mta_tax  tip_amount  \
105063              1.0            0.0         10.0  ...      0.0         0.0   

        tolls_amount  ehail_fee  improvement_surcharge  total_amount  \
105063           0.0        0.0                    0.0          10.0   

        payment_type trip_type  congestion_surcharge   Type  
105063             1       1.0                   0.0  green  

[1 rows x 21 columns]

### fhvhv taxi

In [28]:
fhvhv_taxi = pd.read_csv('fhvhv_tripdata_2019-09.csv', index_col=False)
fhvhv_taxi.head()

hvfhs_license_num dispatching_base_num      pickup_datetime  \
0            HV0003               B02764  2019-09-01 00:10:02   
1            HV0003               B02764  2019-09-01 00:22:37   
2            HV0003               B02764  2019-09-01 00:36:03   
3            HV0005               B02510  2019-09-01 00:04:04   
4            HV0005               B02510  2019-09-01 00:33:15   

      dropoff_datetime  PULocationID  DOLocationID  SR_Flag  
0  2019-09-01 00:17:07           136           169      NaN  
1  2019-09-01 00:31:53           169            18      NaN  
2  2019-09-01 01:04:01            94           198      NaN  
3  2019-09-01 00:26:58           114           112      NaN  
4  2019-09-01 00:41:07           112           112      1.0

In [29]:
col_names = list(fhvhv_taxi.columns)
print(col_names)
print(fhvhv_taxi.shape)
fhvhv_taxi.info()

['hvfhs_license_num', 'dispatching_base_num', 'pickup_datetime', 'dropoff_datetime', 'PULocationID', 'DOLocationID', 'SR_Flag']
(20069234, 7)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20069234 entries, 0 to 20069233
Data columns (total 7 columns):
 #   Column                Dtype  
---  ------                -----  
 0   hvfhs_license_num     object 
 1   dispatching_base_num  object 
 2   pickup_datetime       object 
 3   dropoff_datetime      object 
 4   PULocationID          int64  
 5   DOLocationID          int64  
 6   SR_Flag               float64
dtypes: float64(1), int64(2), object(4)
memory usage: 1.0+ GB


In [30]:
fhvhv_taxi.describe().T

count        mean        std  min   25%    50%    75%  \
PULocationID  20069234.0  138.980040  74.990793  1.0  75.0  140.0  211.0   
DOLocationID  20069234.0  141.972003  77.609570  1.0  76.0  141.0  218.0   
SR_Flag        3235493.0    1.000000   0.000000  1.0   1.0    1.0    1.0   

                max  
PULocationID  265.0  
DOLocationID  265.0  
SR_Flag         1.0

In [31]:
unique_values_for_cols = fhvhv_taxi.nunique()
print(unique_values_for_cols)

hvfhs_license_num             4
dispatching_base_num         32
pickup_datetime         2539957
dropoff_datetime        2540393
PULocationID                262
DOLocationID                264
SR_Flag                       1
dtype: int64


In [32]:
fhvhv_taxi.isnull().sum()

hvfhs_license_num              0
dispatching_base_num           0
pickup_datetime                0
dropoff_datetime               0
PULocationID                   0
DOLocationID                   0
SR_Flag                 16833741
dtype: int64

In [33]:
fhvhv_taxi['SR_Flag'] = fhvhv_taxi['SR_Flag'].fillna(0) # null = non-shared rides, replace with '0' here.

In [34]:
nas=pd.DataFrame(fhvhv_taxi.isnull().sum().sort_values(ascending=False)/len(fhvhv_taxi)*100,columns = ['percent'])
pos = nas['percent'] > 0
nas[pos]

Empty DataFrame
Columns: [percent]
Index: []

• HV0002: Juno
• HV0003: Uber
• HV0004: Via
• HV0005: Lyft

In [35]:
#fhvhv_taxi['Type'] = fhvhv_taxi['Type'][i for i in hvfhs_license_num if i= 'HV0002', ]

values = []

for i in fhvhv_taxi.hvfhs_license_num:
    if i =='HV0002':
        values.append('Juno')
    elif i =='HV0003':
        values.append('Uber')
    elif i =='HV0004':
        values.append('Via')
    elif i =='HV0005':
        values.append('Lyft')


fhvhv_taxi['Type'] = values
fhvhv_taxi.head()

hvfhs_license_num dispatching_base_num      pickup_datetime  \
0            HV0003               B02764  2019-09-01 00:10:02   
1            HV0003               B02764  2019-09-01 00:22:37   
2            HV0003               B02764  2019-09-01 00:36:03   
3            HV0005               B02510  2019-09-01 00:04:04   
4            HV0005               B02510  2019-09-01 00:33:15   

      dropoff_datetime  PULocationID  DOLocationID  SR_Flag  Type  
0  2019-09-01 00:17:07           136           169      0.0  Uber  
1  2019-09-01 00:31:53           169            18      0.0  Uber  
2  2019-09-01 01:04:01            94           198      0.0  Uber  
3  2019-09-01 00:26:58           114           112      0.0  Lyft  
4  2019-09-01 00:41:07           112           112      1.0  Lyft

In [36]:
fhvhv_taxi.nunique()

hvfhs_license_num             4
dispatching_base_num         32
pickup_datetime         2539957
dropoff_datetime        2540393
PULocationID                262
DOLocationID                264
SR_Flag                       2
Type                          4
dtype: int64

In [37]:
duplicateRowsfhvhv = fhvhv_taxi[fhvhv_taxi.duplicated()]
duplicateRowsfhvhv

hvfhs_license_num dispatching_base_num      pickup_datetime  \
4116                HV0003               B02864  2019-09-01 00:59:20   
26855               HV0003               B02764  2019-09-01 00:12:54   
33088               HV0003               B02872  2019-09-01 00:48:24   
299333              HV0003               B02836  2019-09-01 13:42:02   
344603              HV0003               B02395  2019-09-01 14:56:22   
...                    ...                  ...                  ...   
19676678            HV0003               B02875  2019-09-30 08:49:02   
19772693            HV0003               B02765  2019-09-30 13:35:12   
19791539            HV0003               B02872  2019-09-30 13:43:22   
19813984            HV0003               B02864  2019-09-30 14:18:46   
19961247            HV0003               B02883  2019-09-30 19:51:49   

             dropoff_datetime  PULocationID  DOLocationID  SR_Flag  Type  
4116      2019-09-01 01:09:07           174           259      1.0  Uber  
26855     2019-09-01 00:29:33           131           198      1.0  Uber  
33088     2019-09-01 00:54:41            16            16      1.0  Uber  
299333    2019-09-01 13:48:35            48           246      1.0  Uber  
344603    2019-09-01 15:29:32           193           194      1.0  Uber  
...                       ...           ...           ...      ...   ...  
19676678  2019-09-30 08:56:57           168            74      1.0  Uber  
19772693  2019-09-30 13:42:34           239            43      1.0  Uber  
19791539  2019-09-30 13:54:06            92           171      1.0  Uber  
19813984  2019-09-30 14:23:49           244           244      1.0  Uber  
19961247  2019-09-30 19:58:31           140           263      1.0  Uber  

[621 rows x 8 columns]

### fhv Taxi

In [38]:
fhv_taxi = pd.read_csv('fhv_tripdata_2019-09.csv', index_col=False)
fhv_taxi.head()

dispatching_base_num      pickup_datetime     dropoff_datetime  \
0               B00009  2019-09-01 00:35:00  2019-09-01 00:59:00   
1               B00009  2019-09-01 00:48:00  2019-09-01 01:09:00   
2               B00014  2019-09-01 00:16:18  2019-09-02 00:35:37   
3               B00014  2019-09-01 00:55:03  2019-09-01 01:09:35   
4               B00014  2019-09-01 00:13:08  2019-09-02 01:12:31   

   PULocationID  DOLocationID  SR_Flag  
0         264.0         264.0      NaN  
1         264.0         264.0      NaN  
2         264.0         264.0      NaN  
3         264.0         264.0      NaN  
4         264.0         264.0      NaN

In [39]:
col_names = list(fhv_taxi.columns)
print(col_names)
print(fhv_taxi.shape)
fhv_taxi.info()

['dispatching_base_num', 'pickup_datetime', 'dropoff_datetime', 'PULocationID', 'DOLocationID', 'SR_Flag']
(1039487, 6)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1039487 entries, 0 to 1039486
Data columns (total 6 columns):
 #   Column                Non-Null Count    Dtype  
---  ------                --------------    -----  
 0   dispatching_base_num  1039487 non-null  object 
 1   pickup_datetime       1039487 non-null  object 
 2   dropoff_datetime      1039487 non-null  object 
 3   PULocationID          1038935 non-null  float64
 4   DOLocationID          1038935 non-null  float64
 5   SR_Flag               0 non-null        float64
dtypes: float64(3), object(3)
memory usage: 47.6+ MB


In [40]:
fhv_taxi.describe().T

count        mean        std  min    25%    50%    75%  \
PULocationID  1038935.0  225.888643  70.835081  1.0  223.0  264.0  264.0   
DOLocationID  1038935.0  179.432477  88.303258  1.0   94.0  210.0  264.0   
SR_Flag             0.0         NaN        NaN  NaN    NaN    NaN    NaN   

                max  
PULocationID  265.0  
DOLocationID  265.0  
SR_Flag         NaN

In [41]:
unique_values_for_cols = fhv_taxi.nunique()
print(unique_values_for_cols)

dispatching_base_num       419
pickup_datetime         649685
dropoff_datetime        665470
PULocationID               262
DOLocationID               262
SR_Flag                      0
dtype: int64


In [42]:
fhv_taxi.isnull().sum()

dispatching_base_num          0
pickup_datetime               0
dropoff_datetime              0
PULocationID                552
DOLocationID                552
SR_Flag                 1039487
dtype: int64

In [43]:
nas=pd.DataFrame(fhv_taxi.isnull().sum().sort_values(ascending=False)/len(fhv_taxi)*100,columns = ['percent'])
pos = nas['percent'] > 0
nas[pos]

percent
SR_Flag       100.000000
DOLocationID    0.053103
PULocationID    0.053103

In [44]:
fhv_taxi['SR_Flag'] = fhv_taxi['SR_Flag'].fillna(1) # fhv = shared rides, replace with '0' here.

In [45]:
fhv_df = fhv_taxi.dropna(axis = 0, how ='any') 
  
# comparing sizes of data frames 
print("Old data frame length:", len(fhv_taxi), "\nNew data frame length:",  
       len(fhv_df), "\nNumber of rows with at least 1 NA value: ", 
       (len(fhv_taxi)-len(fhv_df)))

Old data frame length: 1039487 
New data frame length: 1038935 
Number of rows with at least 1 NA value:  552


In [46]:
duplicateRowsfhv = fhv_taxi[fhv_taxi.duplicated()]
duplicateRowsfhv

dispatching_base_num      pickup_datetime     dropoff_datetime  \
1332                  B01700  2019-09-01 01:33:00  2019-09-01 01:55:00   
2911                  B01020  2019-09-01 04:00:00  2019-09-01 04:30:00   
4260                  B01537  2019-09-01 06:59:00  2019-09-01 07:13:00   
4409                  B02293  2019-09-01 06:49:23  2019-09-01 07:02:55   
4805                  B01020  2019-09-01 07:30:00  2019-09-01 07:40:00   
...                      ...                  ...                  ...   
1039233               B01537  2019-09-30 23:09:00  2019-09-30 23:23:00   
1039234               B01537  2019-09-30 23:14:00  2019-09-30 23:28:00   
1039235               B01537  2019-09-30 23:29:00  2019-09-30 23:43:00   
1039236               B01537  2019-09-30 23:14:00  2019-09-30 23:28:00   
1039237               B01537  2019-09-30 23:29:00  2019-09-30 23:43:00   

         PULocationID  DOLocationID  SR_Flag  
1332            264.0         264.0      1.0  
2911            264.0         264.0      1.0  
4260            264.0         264.0      1.0  
4409            213.0         244.0      1.0  
4805            264.0         264.0      1.0  
...               ...           ...      ...  
1039233         264.0         264.0      1.0  
1039234         264.0         264.0      1.0  
1039235         264.0         264.0      1.0  
1039236         264.0         264.0      1.0  
1039237         264.0         264.0      1.0  

[7019 rows x 6 columns]